src/tab.rs

```diff
@@ -0,0 +1,295 @@
+use std::collections::HashMap;
+
+use crate::webview::{prompt::PromptDialog, WebView};
+use base::id::WebViewId;
+use serde::{Deserialize, Serialize};
+use servo_url::ServoUrl;
+use webrender_api::units::DeviceIntRect;
+
+/// Tab state
+pub struct Tab {
+    /// Tab WebView id
+    id: WebViewId,
+    /// Tab WebView
+    webview: WebView,
+    /// History
+    history: TabHistory,
+    /// Prompt
+    prompt: Option<PromptDialog>,
+}
+
+impl Tab {
+    /// Create a new tab state.
+    pub fn new(webview: WebView) -> Self {
+        Self {
+            id: webview.webview_id,
+            webview,
+            history: TabHistory {
+                list: Vec::new(),
+                current_idx: 0,
+            },
+            prompt: None,
+        }
+    }
+
+    /// Get tab WebView id.
+    pub fn id(&self) -> WebViewId {
+        self.id
+    }
+
+    /// Get tab WebView.
+    pub fn webview(&self) -> &WebView {
+        &self.webview
+    }
+
+    /// Set tab WebView size.
+    pub fn set_webview_size(&mut self, rect: DeviceIntRect) {
+        self.webview.set_size(rect);
+    }
+
+    /// Get tab history.
+    pub fn history(&self) -> &TabHistory {
+        &self.history
+    }
+
+    /// Set tab history.
+    pub fn set_history(&mut self, list: Vec<ServoUrl>, current_idx: usize) {
+        self.history = TabHistory { list, current_idx };
+    }
+
+    /// Get tab prompt dialog.
+    pub fn prompt(&self) -> Option<&PromptDialog> {
+        self.prompt.as_ref()
+    }
+
+    /// Get tab prompt id.
+    pub fn prompt_id(&self) -> Option<WebViewId> {
+        self.prompt.as_ref().map(|p| p.id())
+    }
+
+    /// Set tab prompt dialog.
+    pub fn set_prompt(&mut self, prompt: PromptDialog) {
+        self.prompt = Some(prompt);
+    }
+
+    /// Remove tab prompt dialog.
+    pub fn remove_prompt(&mut self) -> Option<PromptDialog> {
+        self.prompt.take()
+    }
+
+    /// Check if there is a prompt dialog.
+    pub fn has_prompt(&self) -> bool {
+        self.prompt.is_some()
+    }
+
+    /// Set prompt webview size.
+    pub fn set_prompt_size(&mut self, rect: DeviceIntRect) {
+        if let Some(prompt) = self.prompt.as_mut() {
+            prompt.set_size(rect);
+        }
+    }
+}
+
+/// Tab manager to handle multiple tab in a window.
+pub struct TabManager {
+    /// Current active tab id
+    active_tab_id: Option<WebViewId>,
+    /// Tab webview id -> Tab webview
+    tab_map: HashMap<WebViewId, Tab>,
+    /// Prompt webview id -> Parent tab webview id
+    prompt_tab_map: HashMap<WebViewId, WebViewId>,
```

Comment:
Why does prompt tab still another hashmap here?

Reply:
When the user interacts with the prompt, the constellation will receive a prompt embedder message with the prompt's webview id. We'll need to find the prompt instance using the prompt ID.
If we want to remove the hashmap, then it will need to iterate all tabs to find the prompt.